modules/internal/String.chpl

```diff
@@ -1382,15 +1382,23 @@ module String {
     proc _getView(r:range(?)) where r.idxType == byteIndex {
       if boundsChecking {
         if r.hasLowBound() && (!r.hasHighBound() || r.size > 0) {
-          if r.low:int <= 0 then
-            halt("range out of bounds of string");
+          if r.low:int < 0 then
+            halt("range ", r, " out of bounds for string");
         }
         if r.hasHighBound() && (!r.hasLowBound() || r.size > 0) {
-          if (r.high:int < 0) || (r.high:int > this.len) then
-            halt("range out of bounds of string");
+          // This seems suspicious... why would a range with a high bound
+          // of -1 be in-bounds yet one whose high bound was -2 be out?
+          // It seems as though any bound < 0 or >= len should be OOB.
+          // (This logic pre-dated this PR, though the numbers differed
+          // in the 1-based string/bytes indexing world).
+          // I think that this exists in order to permit the doReplace()
+          // call to work when `find()` returns 0, but this doesn't
+          // seem principled.  See also the similar case in Bytes.chpl.
+          if (r.high:int < -1) || (r.high:int >= this.len) then
+            halt("range ", r, " out of bounds for string with ", this.numBytes, " bytes");
         }
       }
-      const r1 = r[1:r.idxType..this.len:r.idxType];
+      const r1 = r[0:r.idxType..(this.len-1):r.idxType];
       if r1.stridable {
         const ret = r1.low:int..r1.high:int by r1.stride;
```

Comment:
Not related to this PR, but shouldn't that be `r1.alignedLow:int..r1.alignedHigh:int by r1.stride`?

Reply:
```chapel
var s = "123456789abcdef";
var sv = s[1:byteIndex..10:byteIndex by 2 align 2:byteIndex ];
writeln(sv);
```

(on master without this PR) produces the odd numbers, not the even ones and a.

Reply:
I opened #15408 for this..